src/decoder.rs

```diff
@@ -1,3 +1,15 @@
+#[cfg(feature = "std")]
+use std::{collections::HashSet as Set, iter, vec::Vec};
+
+#[cfg(feature = "metal")]
+use core::iter;
+
+#[cfg(feature = "metal")]
+use alloc::{collections::BTreeSet as Set, vec::Vec};
```

Comment:
I'm not familiar with no_std support. Is this best practice, vs always using `alloc`?

Reply:
yes, `std` has many advantages over `no-std` and should always be used when possible. Even when it seems that there is no practical difference, these things could be used by optimizing tools and are relied on in security analysis; finally, things might diverge between `alloc` and `std` further in future, thus maintainability.

Here, it is quite clear, that there is no `HashSet` outside of `std`, thus we have to use `BTreeSet` which, as far as I understand, may have in general inferior performance for the task here.